applogic/src/api/message_list_cubit.rs

```diff
@@ -0,0 +1,214 @@
+// SPDX-FileCopyrightText: 2024 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::{collections::HashMap, pin::pin, sync::Arc};
+
+use flutter_rust_bridge::frb;
+use phnxcoreclient::{
+    store::{Store, StoreEntityId, StoreNotification, StoreOperation},
+    ConversationId, ConversationMessage,
+};
+use tokio::sync::watch;
+use tokio_stream::{Stream, StreamExt};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error};
+
+use crate::{
+    util::{spawn_from_sync, Cubit, CubitCore},
+    StreamSink,
+};
+
+use super::{
+    types::{UiConversationMessage, UiConversationMessageId},
+    user::user_cubit::UserCubitBase,
+};
+
+#[frb(opaque)]
+#[derive(Debug, Default, Clone)]
+pub struct MessageListState {
+    /// Copy-on-write inner ref to make the state cheaply clonable when emitting new state
+    inner: Arc<MessageListStateInner>,
+}
+
+#[frb(ignore)]
+#[derive(Debug, Default)]
+struct MessageListStateInner {
+    /// loaded messages (not all messages in the conversation)
+    messages: Vec<UiConversationMessage>,
+    /// lookup index from message id to index in `messages`
+    message_ids_index: HashMap<UiConversationMessageId, usize>,
+}
+
+impl MessageListState {
```

Comment:
Can you add comments to (at least) the public functions?

Reply:
Good point! Done https://github.com/phnx-im/infra/pull/282/commits/168eb83a2f705d8a464bc38c47843cff4ba69ba7